cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java

```diff
@@ -196,100 +206,122 @@ public final ProgramController run(final Program program, final ProgramOptions o
+                  .split(hConf.get(YarnConfiguration.YARN_APPLICATION_CLASSPATH, ""))))
+                .withApplicationClassPaths(Splitter.on(",").trimResults().split(yarnAppClassPath))
+                .withBundlerClassAcceptor(new HadoopClassExcluder() {
+                  @Override
+                  public boolean accept(String className, URL classUrl, URL classPathUrl) {
+                    // Exclude both hadoop and spark classes.
+                    return super.accept(className, classUrl, classPathUrl)
+                      && !className.startsWith("org.apache.spark.");
+                  }
+                })
+                .withApplicationArguments(
+                  "--" + RunnableOptions.JAR, programJarName,
+                  "--" + RunnableOptions.HADOOP_CONF_FILE, HADOOP_CONF_FILE_NAME,
+                  "--" + RunnableOptions.CDAP_CONF_FILE, CDAP_CONF_FILE_NAME,
+                  "--" + RunnableOptions.APP_SPEC_FILE, APP_SPEC_FILE_NAME,
+                  "--" + RunnableOptions.PROGRAM_OPTIONS, programOptions,
+                  "--" + RunnableOptions.PROGRAM_ID, GSON.toJson(program.getId().toEntityId())
+                );
+
+              TwillController twillController;
+              // Change the context classloader to the combine classloader of this ProgramRunner and
+              // all the classloaders of the dependencies classes so that Twill can trace classes.
+              ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(new CombineClassLoader(
+                AbstractDistributedProgramRunner.this.getClass().getClassLoader(),
+                Iterables.transform(dependencies, new Function<Class<?>, ClassLoader>() {
+                  @Override
+                  public ClassLoader apply(Class<?> input) {
+                    return input.getClassLoader();
+                  }
+                })));
+              try {
+                twillController = twillPreparer.start();
+              } finally {
+                ClassLoaders.setContextClassLoader(oldClassLoader);
+              }
+              return addCleanupListener(twillController, program, tempDir);
+            }
+          });
         }
-      });
+      };
+
+      if (!SecurityUtil.isKerberosEnabled(cConf)) {
+        // simply launch as the current user, if security isn't enabled
+        return callable.call();
+      }
+      // otherwise, execute as the configured user
+      String principal = options.getArguments().getOption(ProgramOptionConstants.PRINCIPAL);
+      String keytabURI = options.getArguments().getOption(ProgramOptionConstants.KEYTAB_URI);
+      ImpersonationInfo impersonationInfo = new ImpersonationInfo(principal, keytabURI);
+      UserGroupInformation impersonatedUGI =
+        new DefaultUGIProvider(cConf, locationFactory).getConfiguredUGI(impersonationInfo);
+      return ImpersonationUtils.doAs(impersonatedUGI, callable);
```

Comment:
instead would it be better if ``DefaultUGIProvider`` returns either a wrapper or a class that extends ``UserGroupInformation``? That way, you don't need a static method like this